### PR TITLE
✍️ Expand the Flax annotated MNIST image classification tutorial, port to the Linen API

### PR DESCRIPTION
**Pull request by 8bitmp3**
Hi @avital 👋 PTAL 🎄

- I annotated the Flax image classification example with the Linen API from the Linen examples (https://github.com/google/flax/blob/master/linen_examples/mnist/train.py) and addressed issue https://github.com/google/flax/issues/659 (porting the old MNIST notebook).
- Format: Markdown, converted using [Jupytext](https://jupytext.readthedocs.io/en/latest/install.html) from a Jupyter notebook (Colab) (to make the review process easier https://github.com/google/flax/issues/659; I have a Jupyter notebook ready).
- I researched the documentation (thanks @avital @mattjj @skye and everyone for awesome docs) and included various references to JAX and Flax guides and API references on RTD to help new users (e.g. Managing state and params, Flax basics, using/splitting PRNG keys (the JAX sharp bits), JAX pytrees).
- I've used the [Google](https://developers.google.com/style), [Microsoft](https://docs.microsoft.com/en-us/style-guide/welcome/), and NumPy style guidelines as much as possible to help with user experience.

LMKWYT 👍 

Example snippets:

> This tutorial uses Flax—a high-performance deep learning library for JAX designed for flexibility—to show you how to construct a simple convolutional neural network (CNN) using the Linen API and train the network for image classification on the MNIST dataset.
> 
> If you're new to JAX, check out [JAX quickstart](https://jax.readthedocs.io/en/latest/notebooks/quickstart.html) and [JAX for the impatient](https://flax.readthedocs.io/en/latest/notebooks/jax_for_the_impatient.html). To learn more about Flax and its Linen API, refer to [Flax basics](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html), [Flax patterns: Managing state and parameters](https://flax.readthedocs.io/en/latest/patterns/state_params.html), and [Linen design principles](https://flax.readthedocs.io/en/latest/design_notes/linen_design_principles.html).

...

> ## Build the Linen model
> 
> Build a convolutional neural network with the Flax Linen API by subclassing [`flax.linen.Module`](https://flax.readthedocs.io/en/latest/flax.linen.html#core-module-abstraction). Because the architecture in this example is relatively simple—you're just stacking layers—you can define the inlined submodules directly within the `__call__` method and wrap it with the `@compact` decorator ([`flax.linen.compact`](https://flax.readthedocs.io/en/latest/flax.linen.html#compact-methods)).

...

> 2. Define a function for your optimizer that takes the model parameters, the learning rate, and a beta argument (for the [Momentum optimizer](http://www.columbia.edu/~nq6/publications/momentum.pdf)):
>   - Choose `Momentum` from the [`flax.optim`](https://flax.readthedocs.io/en/latest/flax.optim.html) package; and
>   - Wrap the model parameters (`params`) with the [`flax.optim.OptimizerDef.create`](https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.OptimizerDef.create) method and initialize with parameter dicts.
> 
> 
> ```python
> def create_optimizer(params, learning_rate, beta):
>   optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
>   optimizer = optimizer_def.create(params)
>   return optimizer
> ```

...

> ## Training and evaluation functions
> 
> 1. Write a training step function that:
> 
>   - Evaluates the neural network given the parameters and a batch of input images with the [`flax.linen.apply`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.apply) method.
>   - Computes the `cross_entropy_loss` loss function.
>   - Evaluates the loss function and its gradient using [`jax.value_and_grad`](https://jax.readthedocs.io/en/latest/jax.html#jax.value_and_grad) (check the [JAX autodiff cookbook](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Evaluate-a-function-and-its-gradient-using-value_and_grad) to learn more).
>   - Applies a [pytree](https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions) of gradients ([`flax.optim.Optimizer.apply_gradient`](https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.Optimizer.apply_gradient)) to the optimizer to update the model's parameters.
>   - Computes the metrics using `compute_metrics` (defined earlier).
> 
>   Use the `@jit` decorator—the entire function is now just-in-time([`jit`](https://jax.readthedocs.io/en/latest/jax.html#jax.jit))-compiled with [XLA](https://www.tensorflow.org/xla) to make your code run faster and more efficiently.
> 
> 
> ```python
> @jax.jit
> def train_step(optimizer, batch):
>   def loss_fn(params):
>     logits = CNN().apply({'params': params}, batch['image'])
>     loss = cross_entropy_loss(logits, batch['label'])
>     return loss, logits
>   grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
>   (_, logits), grad = grad_fn(optimizer.target)
>   optimizer = optimizer.apply_gradient(grad)
>   metrics = compute_metrics(logits, batch['label'])
>   return optimizer, metrics
> ```

...

> 4. Create a model evaluation function that:
> 
>   - Evalues the model on the test set.
>   - Retrieves the evaluation metrics from the device with `jax.device_get`.
>   - Copies the metrics [data stored](https://flax.readthedocs.io/en/latest/design_notes/linen_design_principles.html#how-are-parameters-represented-and-how-do-we-handle-general-differentiable-algorithms-that-update-stateful-variables) in a JAX [pytree]([pytrees](https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions)).
>   - Returns the test loss and accuracy.
> 
> 
> ```python
> def eval_model(model, test_ds):
>   metrics = eval_step(model, test_ds)
>   metrics = jax.device_get(metrics)
>   eval_summary = jax.tree_map(lambda x: x.item(), metrics)
>   return eval_summary['loss'], eval_summary['accuracy']
> ```

...

> ## Initialize the weights and instantiate the optimizer
> 
> 1. Before you start training the model, you need to randomly initialize the weights/parameters.
> 
>   In NumPy, you would usually use the stateful pseudorandom number generators (PRNG). JAX, however, uses an explicit PRNG (refer to [JAX - the sharp bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#JAX-PRNG) for details):
> 
>   - Start by getting two unsigned 32-bit integers with [`jax.random.PRNGKey`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.PRNGKey.html#jax.random.PRNGKey) as one key (`rng`).
>   - Then, split the PRNG to obtain a usable subkey for parameter initialization (`init_rng` below) using [`jax.random.split`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.split.html#jax.random.split). 
> 
>   Note that in JAX and Flax you can have [separate PRNG chains](https://flax.readthedocs.io/en/latest/design_notes/linen_design_principles.html#how-are-parameters-represented-and-how-do-we-handle-general-differentiable-algorithms-that-update-stateful-variables) (with different names, such as `rng` and `init_rng` below) inside `Module`s for different applications.
> 
> 
> ```python
> rng = jax.random.PRNGKey(0)
> rng, init_rng = jax.random.split(rng)
> ```

### CODE REVIEW

File: docs/annotated_mnist.md [hunk 1]
```diff
@@ -1,157 +1,317 @@
...
-@jax.vmap
-def cross_entropy_loss(logits, label):
-  return -logits[label]
+  - Set the initial shape of the kernel (note that JAX and Flax are [row-based](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Model-parameters-&-initialization)); and
+  - Initialize the module parameters of your network (`CNN`) with the [`flax.linen.Module.init`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.init) method using the PRNGKey and parameters (note that the parameters aren't stored with the models).
+
+
+```python
+def get_initial_params(key):
+  init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
```
Comment (avital) [inline docs/annotated_mnist.md:99]: This is an incorrect variable name -- this isn't a shape, it's an array.

File: docs/annotated_mnist.md [hunk 2]
```diff
@@ -1,157 +1,317 @@
...
+
+```python
+def get_datasets():
+  ds_builder = tfds.builder('mnist')
+  ds_builder.download_and_prepare()
+  # Split into training/test sets
+  train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
+  test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
+  # Convert to floating-points
+  train_ds['image'] = jnp.float32(train_ds['image']) / 255.0
```
Comment (avital) [inline docs/annotated_mnist.md:133]: It's common to use this better conversion logic: `(int_val + 0.5) / 256.0`

File: docs/annotated_mnist.md [hunk 3]
```diff
@@ -1,157 +1,317 @@
...
+  test_ds['image'] = jnp.float32(test_ds['image']) / 255.0
+  return train_ds, test_ds
+```
+
+## Training and evaluation functions
+
+1. Write a training step function that:
+
+  - Evaluates the neural network given the parameters and a batch of input images with the [`flax.linen.apply`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.apply) method.
```
Comment (avital) [inline docs/annotated_mnist.md:142]: It's not `flax.linen.apply`. It's `Module.apply`.

File: docs/annotated_mnist.md [hunk 4]
```diff
@@ -1,157 +1,317 @@
...
+  grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+  (_, logits), grad = grad_fn(optimizer.target)
   optimizer = optimizer.apply_gradient(grad)
-  return optimizer
+  metrics = compute_metrics(logits, batch['label'])
+  return optimizer, metrics
 ```
 
-Making model predictions is as simple as calling `model(input)`:
+2. Create a [`jit`](https://jax.readthedocs.io/en/latest/jax.html#jax.jit)-compiled function that evaluates the model on the test set using [`flax.linen.apply`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.apply):
```
Comment (avital) [inline docs/annotated_mnist.md:165]: Same here, it's not `flax.linen.apply`.

File: docs/annotated_mnist.md [hunk 5]
```diff
@@ -1,157 +1,317 @@
...
+  - Evalues the model on the test set.
+  - Retrieves the evaluation metrics from the device with `jax.device_get`.
+  - Copies the metrics [data stored](https://flax.readthedocs.io/en/latest/design_notes/linen_design_principles.html#how-are-parameters-represented-and-how-do-we-handle-general-differentiable-algorithms-that-update-stateful-variables) in a JAX [pytree](https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions).
+  - Returns the test loss and accuracy.
 
-```py
-def train():
+
+```python
+def eval_model(model, test_ds):
```
Comment (avital) [inline docs/annotated_mnist.md:218]: The argument should be `params`, not `model`.

**Review by avital — COMMENTED** [review]
Hi @8bitmp3, thanks for this contribution! I like it -- I added a bunch of comments. I'm not sure though whether this should be  our official annotated MNIST, as it's somewhat verbose. I think we should have a straight port of the old annotated MNIST to the new code. Most people are happy to look through the reference docs for functions that they don't understand, but the converse is less true -- if people open a verbose doc, then they are likely to simply close the tab and not read through...

That being said, I think this is extremely valuable content! Maybe you'd like to put this on a blog post somewhere, and then we can add a link to it from the top of a ported annotated MNIST that is short? What do you think? I could see a few different options being the right one here, so I'm just sharing my opinion. Happy to hear other perspectives.

**Comment by 8bitmp3** [comment]
> can add a link to it from the top of a ported annotated MNIST that is short? What do you think? I could see a few different options being the right one here, so I'm just sharing my opinion. Happy to hear other perspectives.

Sounds good 👍 Let's have a straight port (not too simple though, as it's not better than having a `.py` file) + a separate deep dive for the curious (maybe have both Markdown files in the same directory to future-proof it). Coming form TF and PT, I found it challenging to learn the nuts and bolts of Linen at first with information being in different places—in Slides presentations, various Flax (the Basics) and JAX (RNGs, Sharp Bits, etc) guides on different pages. That's why I took a lot of notes from the docs and incorporated them into this longer-form notebook.

> Most people are happy to look through the reference docs for functions that they don't understand, but the converse is less true -- if people open a verbose doc, then they are likely to simply close the tab and not read through...

Understood. Personally—and that applies to PT, LightGBM and other systems—I think digging through API docs is not for the impatient, especially if you're a new deep learning user who's curious about this new thing is building and is not aware of some of the best practices (i.e. googling effectively 🙃). That's why I tend to include all the reference docs to help with UX. A bunch of stuff is also currently explained in detail in various Flax/JAX guides (I noticed some Googlers say: read the Sharp Bits a few times)—and sometimes the API docs have better examples in docstrings—so I had to reference them to save time for others. For instance, TF.org auto-link APIs in Jupyter notebooks/Markdown files after the build—which saves users time—but we have to do it by hand here.

File: docs/annotated_mnist.md (hunk 5 above)
Comment (8bitmp3) [reply docs/annotated_mnist.md:218]: The code is from our `linen_examples`:

Source: https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L131

```py
def eval_model(model, test_ds):
  metrics = eval_step(model, test_ds)
  metrics = jax.device_get(metrics)
  summary = jax.tree_map(lambda x: x.item(), metrics)
  return summary['loss'], summary['accuracy']
```

Let me know if I should still change this

File: docs/annotated_mnist.md (hunk 3 above)
Comment (8bitmp3) [reply docs/annotated_mnist.md:142]: Thanks for the catch! Much appreciated. (This is why I personally think having direct links to API docs is super useful in tutorials—I couldn't find the  right `apply` given my inexperience with Linen.)

File: docs/annotated_mnist.md (hunk 2 above)
Comment (8bitmp3) [reply docs/annotated_mnist.md:133]: Cool. Should we update the MNIST example in `linen_examples`? The code comes directly from there (that's how I started learning Linen 🙂).

https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L138

```py
def get_datasets():
  """Load MNIST train and test datasets into memory."""
  ds_builder = tfds.builder('mnist')
  ds_builder.download_and_prepare()
  train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
  test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
  train_ds['image'] = jnp.float32(train_ds['image']) / 255.
  test_ds['image'] = jnp.float32(test_ds['image']) / 255.
  return train_ds, test_ds
```

File: docs/annotated_mnist.md (hunk 1 above)
Comment (8bitmp3) [reply docs/annotated_mnist.md:99]: OK! 👍 The code came from our Flax `linen_examples` - maybe we should also update the MNIST example there. WDYT?

https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L54

```py
def get_initial_params(key):
  init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
  initial_params = CNN().init(key, init_shape)['params']
  return initial_params
```

File: docs/annotated_mnist.md [hunk 6]
```diff
@@ -1,157 +1,317 @@
...
-```py
-@jax.vmap
-def cross_entropy_loss(logits, label):
-  return -logits[label]
+  - Set the initial shape of the kernel (note that JAX and Flax are [row-based](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Model-parameters-&-initialization)); and
+  - Initialize the module parameters of your network (`CNN`) with the [`flax.linen.Module.init`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.init) method using the PRNGKey and parameters (note that the parameters aren't stored with the models).
+
+
+```python
+def get_initial_params(key):
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/annotated_mnist.md:98]: This is great feedback, still learning Linen 👍 The code came from our Flax `linen_examples` - maybe we should also update the MNIST example there. WDYT?

https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L54

File: docs/annotated_mnist.md [hunk 7]
```diff
@@ -1,157 +1,317 @@
...
+## Build the Linen model
+
+Build a convolutional neural network with the Flax Linen API by subclassing [`flax.linen.Module`](https://flax.readthedocs.io/en/latest/flax.linen.html#core-module-abstraction). Because the architecture in this example is relatively simple—you're just stacking layers—you can define the inlined submodules directly within the `__call__` method and wrap it with the `@compact` decorator ([`flax.linen.compact`](https://flax.readthedocs.io/en/latest/flax.linen.html#compact-methods)).
+
+
+```python
+class CNN(nn.Module):
+
+  @nn.compact
+  # Provide a constructor to register a new parameter
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/annotated_mnist.md:50]: @avital It must be a "thinking out loud comment" when learning about Linen Module that made its way into the `# comment` 🙂

**Comment by 8bitmp3** [comment]
Thanks for the awesome feedback re: code. I'm learning lots! As I mentioned in the replies, the code came from our Flax `linen_examples` - maybe we should also update the MNIST example there. WDYT?

Source: https://github.com/google/flax/blob/master/linen_examples/mnist/train.py

And HNY 🥳!

File: docs/annotated_mnist.md [hunk 8]
```diff
@@ -1,157 +1,317 @@
...
-automatically vectorizes code efficiently to run on entire
-batches.
+3. Create a function for parameter initialization:
 
-```py
-@jax.vmap
-def cross_entropy_loss(logits, label):
-  return -logits[label]
+  - Set the initial shape of the kernel (note that JAX and Flax are [row-based](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Model-parameters-&-initialization)); and
+  - Initialize the module parameters of your network (`CNN`) with the [`flax.linen.Module.init`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.init) method using the PRNGKey and parameters (note that the parameters aren't stored with the models).
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/annotated_mnist.md:94]: Done

File: docs/annotated_mnist.md (hunk 3 above)
Comment (8bitmp3) [reply docs/annotated_mnist.md:142]: Corrected it in 2 places. Thank you again

**Comment by 8bitmp3** [comment]
@avital PTAL. A lighter and updated version with a small intro. It also includes shorter API names with embedded links to make it easier to new users to find the relevant API docs. I strongly believe that links to various guides in different steps should be included—they've personally helped me a lot to get to know Linen and JAX better and I hope new adopters will appreciate them too 👍 Users should appreciate well-documented stuff, hopefully.

As for the code comments, I answered them separately—I can't change anything there until we decide what to do with the `linen_examples/mnist` example, since the code comes from there and I think we should have consistent approaches here and there 🙂

Thank you again 🙏

**Comment by avital** [comment]
Hi @8bitmp3 -- sorry for such a slow review cycle, my kids' schools being closed has shaken up my productivity... This looks really good, thanks for taking the time to do it. I'm happy to merge it, but I think it should live in RTD/ rather than a README. So why don't I merge this in first, and then maybe would you like to file a subsequent PR moving this file into RTD and making sure it appears in a good place in the ToC?

**Comment by avital** [comment]
@8bitmp3 just circling back to this -- I don't think many people find this tutorial because it's in the codebase and not in RTD. Do you have time to move it into RTD and make sure it appears somewhere nice in the ToC?

**Pull request merged**